### PR TITLE
Fix #379 -- Add logo to event organizers

### LINKED ISSUE #379: Add logo to event organizers
What's about adding a "logo" attribute to event organizers (similar to events) which is displayed on the front page of each event organizer?

### PR DESCRIPTION
**Pull request by b-lawrence**
Fixes #379

With this solution, general organizer updates, and settings happen on a single page(single view), but still uses different forms. To improve this, even more, we can have organizer settings tabs similar to events with separate views, just so that things don't start to look messy when `organizer settings` takes on a little more. 

Awaiting feedback :)

### CODE REVIEW

**Comment by b-lawrence** [comment]
Looks like I need to add in some tests here. 😄

**Comment by b-lawrence** [comment]
@raphaelm @jfwiebe What do you guys think? 😄

File: src/pretix/control/forms/organizer.py [hunk 1]
```diff
@@ -25,9 +26,20 @@ def clean_slug(self):
...
         return self.instance.slug
+
+
+class OrganizerSettingsForm(SettingsForm):
+
+    logo_image = ExtFileField(
+        label=_('Logo image'),
+        ext_whitelist=(".png", ".jpg", ".svg", ".gif", ".jpeg"),
+        required=False,
+        help_text=_('Add a logo to your organizer page.')  # TODO: Use a better desc.
```
Comment (raphaelm) [inline src/pretix/control/forms/organizer.py]: How about something similar to the event form, e.g:

```
        help_text=_('If you provide a logo image, we will by default not show your organization name '
                    'in the page header. We will show your logo with a maximal height of 120 pixels.')
```

File: src/pretix/control/templates/pretixcontrol/organizers/detail.html [hunk 2]
```diff
@@ -7,7 +7,7 @@
         {% blocktrans with name=organizer.name %}Organizer: {{ name }}{% endblocktrans %}
         <a href="{% url "control:organizer.edit" organizer=organizer.slug %}"
            class="btn btn-default">
-            <span class="fa fa-edit"></span>
+            <span class="fa fa-sliders"></span>
```
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/organizers/detail.html]: May I ask why you changed this? We use ``fa-edit`` a lot for editing (and ``fa-wrench`` for event configuration) and I think it is better to stay consistent than to introduce ``fa-sliders``.

File: src/pretix/control/views/organizer.py [hunk 3]
```diff
@@ -195,13 +198,50 @@ class OrganizerUpdate(OrganizerPermissionRequiredMixin, UpdateView):
     model = Organizer
     form_class = OrganizerUpdateForm
     template_name = 'pretixcontrol/organizers/edit.html'
-    permission = None
+    permission = 'can_change_settings'
```
Comment (raphaelm) [inline src/pretix/control/views/organizer.py]: This is useless, as ``OrganizerPermission`` currently has no ``can_change_settings`` attribute. (The permission system is not that well-polished, we know...)

File: src/pretix/presale/templates/pretixpresale/event/base.html [hunk 4]
```diff
@@ -22,16 +22,16 @@
...
-        <h1 class="pull-left">
+        <div class="pull-left">
             {% if event_logo %}
                 <a href="{% eventurl event "presale:event.index" %}" title="{{ event.name }}">
                     <img src="{{ event_logo|thumbnail_url:'logo' }}" alt="{{ event.name }}" class="event-logo" />
                 </a>
             {% else %}
-                <a href="{% eventurl event "presale:event.index" %}">{{ event.name }}</a>
+                <h1><a href="{% eventurl event "presale:event.index" %}">{{ event.name }}</a></h1>
                 <small>{{ event.get_date_range_display }}</small>
```
Comment (raphaelm) [inline src/pretix/presale/templates/pretixpresale/event/base.html]: Was this intended? Before, the ``<small>`` was a sub-element of the ``<h1>``. Let's keep it this way for now. (It doesn't look that bad in the line below, but we'd need to adjust the language switcher then.)

File: src/pretix/control/forms/organizer.py [hunk 5]
```diff
@@ -25,9 +26,20 @@ def clean_slug(self):
...
         super().__init__(*args, **kwargs)
         self.fields['slug'].widget.attrs['readonly'] = 'readonly'
 
     def clean_slug(self):
         return self.instance.slug
+
+
+class OrganizerSettingsForm(SettingsForm):
+
+    logo_image = ExtFileField(
```
Comment (raphaelm) [inline src/pretix/control/forms/organizer.py]: Here's a subtle thing that I also did not notice during review and that caught me by surprise when I tested your changes: Our settings key-value store is hierarchical. By this I mean that if you try to access a setting of an event and the setting is not set for that event, the setting of the organizer will be accessed instead. If that does not exist, the global setting will be taken and only if this also does not exist, the hardcoded default will be used.

Currently, we rarely use this hierarchy, as it is hard to represent it in the user interface and normally it does not gain any attention as the different levels normally use different keys. Here, however, the effect is that if you set the organizer logo, all events without a logo of their own will inherit the logo. I believe this is not what a user expects, or at least I did not expect that. The fix is easy: Call this thing ``organizer_logo_image`` instead :wink:

Comment (b-lawrence) [inline src/pretix/control/forms/organizer.py]: Interesting! I do see that happening in `SettingsProxy`. Thanks for pointing that out. 👍

File: src/pretix/control/forms/organizer.py (hunk 1 above)
Comment (b-lawrence) [inline src/pretix/control/forms/organizer.py]: Thought as much! Just wanted to be sure.

File: src/pretix/control/forms/organizer.py (hunk 5 above)
Comment (b-lawrence) [inline src/pretix/control/forms/organizer.py]: The way you handle `settings` in the app is pretty neat if you ask me 😄

Comment (raphaelm) [inline src/pretix/control/forms/organizer.py]: Aww, thanks a lot! :)

File: src/pretix/control/templates/pretixcontrol/organizers/detail.html (hunk 2 above)
Comment (b-lawrence) [inline src/pretix/control/templates/pretixcontrol/organizers/detail.html]: Sorry I didn't raise a flag before changing this. My bad!

Bringing in organizer `Display settings` into the general organizer's edit page; I was thinking of an icon that would intuitively tell the user that he/she can do more that just general edits. Something `settlings-like`. However, now that I know why it must remain that way, I'll roll this back. :)

File: src/pretix/control/views/organizer.py (hunk 3 above)
Comment (b-lawrence) [inline src/pretix/control/views/organizer.py]: 🙈

File: src/pretix/control/templates/pretixcontrol/organizers/detail.html (hunk 2 above)
Comment (raphaelm) [inline src/pretix/control/templates/pretixcontrol/organizers/detail.html]: Its not that it *must* or that we don't like changes, its just that I value consistency very highly :wink:

File: src/pretix/presale/templates/pretixpresale/event/base.html (hunk 4 above)
Comment (b-lawrence) [inline src/pretix/presale/templates/pretixpresale/event/base.html]: Surely this wasn't intended. I must have overlooked it or something.

File: src/pretix/control/templates/pretixcontrol/organizers/detail.html (hunk 2 above)
Comment (b-lawrence) [inline src/pretix/control/templates/pretixcontrol/organizers/detail.html]: I do too. 👍

**Comment by b-lawrence** [comment]
Nice! 🙌  Will have tests in soon enough.

**Comment by raphaelm** [comment]
Hi @b-lawrence, everything going alright? Don't want to build up pressure, just ask if there are any problems concerning the tests or if you just haven't found time so far. :)

**Comment by b-lawrence** [comment]
Hi @raphaelm, Yeah! Everything OK. Sorry, I had to run off on some other project, thus, little time for my open source projects. I have started with the tests now and should have it in by tomorrow. ;)

Cheers,

**Comment by raphaelm** [comment]
No worries, take your time! I know well enough how this is.

**Comment by raphaelm** [comment]
Hey @b-lawrence, everything alright on your side? I'll probably need this feature sometime soon, but don't feel stressed by that, if you don't have time to finish this (and maybe rebase it against current master), I'll just finish it up myself, but I didn't want to do that without prior notice :wink:

**Comment by raphaelm** [comment]
Since this was basically done, I took the liberty of rebasing this against current master and merging it, so we can continue working in this area :) Should you want to add the unit tests, please open a separate pull request. Thank you for this contribution, it's well appreciated! :)

**Pull request merged**